include/exadg/operators/adaptive_mesh_refinement.h

```diff
@@ -0,0 +1,163 @@
+/*  ______________________________________________________________________
+ *
+ *  ExaDG - High-Order Discontinuous Galerkin for the Exa-Scale
+ *
+ *  Copyright (C) 2021 by the ExaDG authors
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *  ______________________________________________________________________
+ *
+ * Source: https://github.com/MeltPoolDG/MeltPoolDG
+ * Author: Peter Munch, Magdalena Schreter, TUM, December 2020
+ */
+
+#ifndef INCLUDE_EXADG_OPERATORS_ADAPTIVE_MESH_REFINEMENT_H_
+#define INCLUDE_EXADG_OPERATORS_ADAPTIVE_MESH_REFINEMENT_H_
+
+// deal.II
+#include <deal.II/base/quadrature_lib.h>
+#include <deal.II/distributed/grid_refinement.h>
+#include <deal.II/distributed/solution_transfer.h>
+#include <deal.II/dofs/dof_tools.h>
+#include <deal.II/grid/grid_refinement.h>
+#include <deal.II/numerics/error_estimator.h>
+#include <deal.II/numerics/solution_transfer.h>
+
+
+namespace ExaDG
+{
+struct AdaptiveMeshRefinementData
+{
+  unsigned int trigger_every_n_time_steps = 1;
+
+  int  maximum_refinement_level = 10;
+  int  minimum_refinement_level = 0;
+  bool preserve_boundary_cells  = false;
+
+  double fraction_of_cells_to_be_refined   = 0.0;
+  double fraction_of_cells_to_be_coarsened = 0.0;
+};
+
+/**
+ * Check if adaptive mesh refinement should be triggered depending on the time_step_number.
+ */
+inline bool
+trigger_coarsening_and_refinement_now(const AdaptiveMeshRefinementData & amr_data,
+                                      const unsigned int                 time_step_number)
+{
+  return ((time_step_number == 0) or not(time_step_number % amr_data.trigger_every_n_time_steps));
```

Comment:
For now: We should not pass amr_data to this function in my opinion, but only the relevant parameter. We might have different AMRData structs in the future, while this function could then no longer be used even though the functionality is actually the same.

For the future: Moreover, we should actually realize such functionality generically in ExaDG (i.e. independently of AMR), since this appears for many different problems (e.g. postprocessing, output of solver information, update of preconditioners). As you have written, let's consider this a future task.